File: HPWorkpath_v1.6.3/Samples/ExampleAPIServices/source/MassStorageSample/src/com/hp/workpath/sample/massstoragesample/task/CreateFileTask.java
// Copyright 2018 HP Development Company, L.P.
// SPDX-License-Identifier: MIT
package com.hp.workpath.sample.massstoragesample.task;

import android.os.Handler;
import android.os.Looper;

import com.hp.workpath.api.massstorage.CustomerDataFile;
import com.hp.workpath.api.massstorage.CustomerDataFileUtils;
import com.hp.workpath.sample.massstoragesample.Logger;
import com.hp.workpath.sample.massstoragesample.MainActivity;

import java.io.BufferedOutputStream;
import java.io.IOException;
import java.io.OutputStream;
import java.lang.ref.WeakReference;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class CreateFileTask {

    private final WeakReference<MainActivity> mContextRef;
    private final CustomerDataFile mCustomerDataFile;
    private final String mContent;

    private ExecutorService executor = Executors.newSingleThreadExecutor();
    private Handler handler = new Handler(Looper.getMainLooper());

    public CreateFileTask(MainActivity context, CustomerDataFile customerDataFile, String content) {
        this.mContextRef = new WeakReference<>(context);
        this.mCustomerDataFile = customerDataFile;
        this.mContent = content;
    }

    public void taskExecute() {
        try {
            executor.execute(new Runnable() {
                @Override
                public void run() {
                    try {
                        mCustomerDataFile.createNewFile();

                        OutputStream fos = null;
                        BufferedOutputStream bos = null;
                        try {
                            fos = CustomerDataFileUtils.openOutputStream(mContextRef.get(), mCustomerDataFile);
                            bos = new BufferedOutputStream(fos);
                            bos.write(mContent.getBytes());
                            bos.flush();
                            onPostExecute(true);
                        } catch (Throwable t) {
                            Logger.showResult(mContextRef.get(), "File could not be created. " + t.getMessage());
                        } finally {
                            if (bos != null) {
                                try {
                                    bos.close();
                                } catch (IOException e) { // ignore
                                }
                            }
                            if (fos != null) {
                                try {
                                    fos.close();
                                } catch (IOException e) { // ignore
                                }
                            }
                        }

                    } catch (Throwable t) {
                        Logger.showResult(mContextRef.get(), "File could not be created. " + t.getMessage());
                        executor.shutdown();
                        onPostExecute(false);
                    }
                }
            });
        } catch (Exception e) {
            Logger.showResult(mContextRef.get(), "File could not be created. " + e.getMessage());
            onPostExecute(false);
            executor.shutdown();
        }
    }

    private void onPostExecute(Boolean result) {
        handler.post(new Runnable() {
            @Override
            public void run() {
                mContextRef.get().enableButton(true);
                if (result) {
                    Logger.showResult(mContextRef.get(), "File created");
                    mContextRef.get().displayFileList(mCustomerDataFile.getParentFile());
                }
            }
        });
    }
}

File: HPWorkpath_v1.6.3/Samples_Kotlin/ExampleAPIServices/source/MassStorageSample/src/com/hp/workpath/sample/massstoragesample/task/CreateFileTask.kt
// Copyright 2018 HP Development Company, L.P.
// SPDX-License-Identifier: MIT
package com.hp.workpath.sample.massstoragesample.task

import com.hp.workpath.api.massstorage.CustomerDataFile
import com.hp.workpath.api.massstorage.CustomerDataFileUtils
import com.hp.workpath.sample.massstoragesample.Logger
import com.hp.workpath.sample.massstoragesample.MainActivity
import java.io.BufferedOutputStream
import java.io.IOException
import java.io.OutputStream
import java.lang.ref.WeakReference

class CreateFileTask(context: MainActivity, customerDataFile: CustomerDataFile, content: String) {
    private val mContextRef: WeakReference<MainActivity> = WeakReference(context)
    private val mCustomerDataFile: CustomerDataFile = customerDataFile
    private val mContent: String = content
    private var mThrowable: Throwable? = null

    suspend fun execute() {
        var isFileCreated = false
        try {
            mCustomerDataFile.createNewFile()

            var fos: OutputStream? = null
            var bos: BufferedOutputStream? = null
            try {
                fos = CustomerDataFileUtils.openOutputStream(mContextRef.get(), mCustomerDataFile)
                bos = BufferedOutputStream(fos)
                bos.write(mContent.toByteArray())
                bos.flush()
                isFileCreated = true
            } catch (t: Throwable) {
                mThrowable = t
            } finally {
                if (bos != null) {
                    try {
                        bos.close()
                    } catch (e: IOException) { // ignore
                    }
                }
                if (fos != null) {
                    try {
                        fos.close()
                    } catch (e: IOException) { // ignore
                    }
                }
            }
        } catch (t: Throwable) {
            mThrowable = t
        }
        onPostExecute(isFileCreated)
    }

    private suspend fun onPostExecute(result: Boolean) {
        mContextRef.get()?.run {
            runOnUiThread {
                enableButton(true);
                if (result) {
                    Logger.showResult(this, "File created")
                    displayFileList(mCustomerDataFile.parentFile)
                } else {
                    Logger.showResult(this, "File could not be created. ${mThrowable?.message}")
                }
            }
        }
    }
}